README.md
# gpt-coder README

This is the README for your extension "gpt-coder". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: Enable/disable this extension.
* `myExtension.thing`: Set to `blah` to do something.

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

---

## Working with Markdown

You can author your README using Visual Studio Code.  Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux)
* Toggle preview (`Shift+Cmd+V` on macOS or `Shift+Ctrl+V` on Windows and Linux)
* Press `Ctrl+Space` (Windows, Linux, macOS) to see a list of Markdown snippets

## For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "gpt-coder" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

media/main.js

// This script will be run within the webview itself
// It cannot access the main VS Code APIs directly.
(function () {
    const vscode = acquireVsCodeApi();

    const oldState = vscode.getState() || { colors: [] };

    /** @type {Array<{ value: string }>} */
    let colors = oldState.colors;

    document.querySelector('#get-query').addEventListener('click', () => {
        const queryLanguage = document.getElementById("query-language").value;
        const queryPrompt = document.getElementById("query-prompt").value;   
        if(!queryLanguage.length){
            showError("Please enter target language");
        } 
        if(!queryPrompt.length){
            showError("Please enter a prompt");
        } 
        getQuery(queryLanguage, queryPrompt);
       
    });


    document.querySelector('#explain-query-btn').addEventListener('click', () => {
        const explainLanguage = document.getElementById("explain-language").value;
        const query = document.getElementById("explain-query").value;   
        if(!explainLanguage.length){
            showError("Please enter target language");
        } 
        if(!query.length){
            showError("Please enter a query");
        } 
        getQueryExplanation(explainLanguage, query);
       
    });

    document.querySelector('#translate-query').addEventListener('click', () => {
        const queryLanguage1 = document.getElementById("query-language-1").value;
        const queryLanguage2 = document.getElementById("query-language-2").value;
        const queryPrompt = document.getElementById("query-prompt-2").value;   

        if(!queryLanguage1.length){
            showError("Please enter source lanuage");
        }
        if(!queryLanguage2.length){
            showError("Please enter target lanuage");
        }
        if(!queryPrompt.length){
            showError("Please enter a query");
        } 
        translateQuery(queryLanguage1, queryLanguage2, queryPrompt);
       
    });

    function translateQuery(queryLanguage1, queryLanguage2, queryPrompt) {
        vscode.postMessage({ type: 'queryTranslate', language1: queryLanguage1, language2: queryLanguage2, prompt: queryPrompt});
    }

    function getQuery(queryLanguage, queryPrompt) {
        vscode.postMessage({ type: 'getQuery', language: queryLanguage,  prompt: queryPrompt});
    }

    function showError(error) {
        vscode.postMessage({ type: 'showError', value: error });
    }

    function getQueryExplanation(explainLanguage, query){
        vscode.postMessage({ type: 'explainQuery', language: explainLanguage,  query: query});
    }


}());




utils/email.js
const { window }  = require('vscode');
const checkEmail = async (context) => {
    const email = await context.globalState.get("email");

    // If the email has not been saved, ask the user for their email
    if (!email) {
        const newEmail = await window.showInputBox({
            title: "Code-GPT",
            placeHolder: "📧 Enter your email",
            prompt: "To protect us from unlimited use. We won't spam your email, promise. 🤝",
            validateInput: (value) => {
                const emailRegex =
                    /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

                if (!emailRegex.test(value)) {
                    return "Error: Invalid email. ☹";
                }

                return null;
            },
        });

        await context.globalState.update("email", newEmail);
    }
};

module.exports = {
    checkEmail
}


utils/comments.js
const singleLineCommentMap = {
    abap: "--",
    bat: "::",
    bibtex: "%",
    clojure: ";",
    coffeescript: "#",
    c: "//",
    cpp: "//",
    csharp: "//",
    "cuda-cpp": "//",
    diff: "//",
    dockerfile: "#",
    fsharp: "//",
    "git-commit": "#",
    "git-rebase": "#",
    go: "//",
    groovy: "//",
    ini: ";",
    java: "//",
    javascript: "//",
    javascriptreact: "//",
    json: "//",
    jsonc: "//",
    latex: "%",
    less: "/*",
    lua: "--",
    makefile: "#",
    objectivec: "//",
    objectivecpp: "//",
    perl: "#",
    perl6: "#",
    php: "//",
    plaintext: "//",
    powershell: "#",
    jade: "//",
    pug: "//",
    python: "#",
    r: "#",
    razor: "@*",
    ruby: "#",
    rust: "//",
    sass: "//",
    shaderlab: "//",
    shellscript: "#",
    sql: "--",
    slim: "//",
    stylus: "//",
    swift: "//",
    typescript: "//",
    typescriptreact: "//",
    tex: "%",
    vb: "'",
    yaml: "#",
};

const multiLineCommentMap = {
    css: ["/*", "*/"],
    handlebars: ["{{!--", "--}}"],
    html: ["<!--", "-->"],
    markdown: ["<!--", "-->"],
    scss: ["/*", "*/"],
    xml: ["<!--", "-->"],
    xsl: ["<!--", "-->"],
};

const styleAsComment = (text, lang) => {
    const singleLineComment = singleLineCommentMap[lang];
    const multiLineComment = multiLineCommentMap[lang];

    if (!singleLineComment && !multiLineComment) return text;

    if (singleLineComment) return `${singleLineComment} ${text}`;
    if (multiLineComment) return `${multiLineComment[0]} ${text} ${multiLineComment[1]}`;
};

const languageSupportsComments = (lang) => {
    return !!singleLineCommentMap[lang] || !!multiLineCommentMap[lang];
};

module.exports = {
    styleAsComment,
    languageSupportsComments
}


commands/query_commands.js
const { Configuration, OpenAIApi } = require("openai");
const { displayError } = require('./utils/errors');
const { ExtensionContext, window, commands } = require('vscode');
const { styleAsComment } = require('./../utils/comments');


const key1 = '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'
const key2 = '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'

const config = new Configuration({
	apiKey: key1,
});

const openai = new OpenAIApi(config);

async function getSQLQueryFromSelection() {
    // The code you place here will be executed every time your command is executed
    const editor = window.activeTextEditor;
    if (!editor) return displayError("noEditor");

    // Aquire the selected text
    const selection = editor.selection;
    const selectedText = editor.document.getText(selection);
    if (!selectedText.length) return displayError("noSelection");

    const lang = editor.document.languageId;

    const prompt = `generate sql query as per following description :\n ${selectedText}`;
    // Send the selected code and email to the API
    try {
        const response = await openai.createCompletion({
            model: "text-davinci-003",
            prompt: `${prompt}`,
            temperature: 0,
            max_tokens: 1000,
            top_p: 1,
            frequency_penalty: 0.5,
            presence_penalty: 0,
        });

        if (!response.data) return displayError("serverSentNothing");

        // Get the output from the API response
        const output = response.data.choices[0].text.trim();

        // Wrap the output in a comment
        const comment = styleAsComment(output, lang);

        // Replace the selected text with the output
        editor.edit((editBuilder) => {
            editBuilder.insert(selection.end, `\n${comment}`);
        });

        // Show the output in a message
        window.showInformationMessage(output)

    } catch (error) {
        console.error(error);
        return displayError("unableToConnect");
    }

    // Display a message box to the user
    vscode.window.showInformationMessage('Hello World from gpt-coder!');
}

module.exports = {
    getSQLQueryFromSelection
}

extension.js
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require('vscode');
const { Configuration, OpenAIApi } = require("openai");
const { displayError } = require('./utils/errors');
const { ExtensionContext, window, commands } = require('vscode');
const { styleAsComment } = require('./utils/comments');

const key1 = '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'
const key2 = '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'

const config = new Configuration({
	apiKey: key1,
});

const openai = new OpenAIApi(config);

/**
 * @param {vscode.ExtensionContext} context
 */

function activate(context) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('Congratulations, your extension "gpt-coder" is now active!');

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with  registerCommand
	// The commandId parameter must match the command field in package.json


	context.subscriptions.push(vscode.commands.registerCommand('gpt-coder.getQuery', async function () {
		// The code you place here will be executed every time your command is executed
		const editor = window.activeTextEditor;
		if (!editor) return displayError("noEditor");

		const language = await vscode.window.showInputBox({
			placeHolder: "Language",
			prompt: "Generate a query for a language"
		  });
		  
		if (!language.length) return displayError("Please enter target language");
		
		const prompt = await vscode.window.showInputBox({
			placeHolder: "Query Explanation",
			prompt: "Generate a query for"
		  });

		if (!prompt.length) return displayError("Please enter a prompt to generate a query");

		// Aquire the selected text
		const selection = editor.selection;
		// const selectedText = editor.document.getText(selection);
		

		const lang = editor.document.languageId;

		try {
			const comment = styleAsComment(await getChatGPTAnswer(getPromptForQuery(prompt,language)), lang);
			// Replace the selected text with the output
			editor.edit((editBuilder) => {
				editBuilder.insert(selection.end, `\n${comment}`);
			});

			// Show the output in a message
			window.showInformationMessage(comment)

		} catch (error) {
			console.error(error);
			return displayError("unableToConnect");
		}

	}));


	context.subscriptions.push(vscode.commands.registerCommand('gpt-coder.explainQuery', async function () {
		// The code you place here will be executed every time your command is executed
		const editor = window.activeTextEditor;
		if (!editor) return displayError("noEditor");

		const language = await vscode.window.showInputBox({
			placeHolder: "Language",
			prompt: "Explain a query for a language"
		  });
		  
		if (!language.length) return displayError("Please enter the query language");
		
		const query = await vscode.window.showInputBox({
			placeHolder: "Query for explanation",
			prompt: "Query"
		  });

		if (!query.length) return displayError("Please enter a query");

		// Aquire the selected text
		const selection = editor.selection;
		// const selectedText = editor.document.getText(selection);
		

		const lang = editor.document.languageId;

		try {
			const comment = styleAsComment(await getChatGPTAnswer(getPromptForQueryExplain(language, query)), lang);
			// Replace the selected text with the output
			editor.edit((editBuilder) => {
				editBuilder.insert(selection.end, `\n${comment}`);
			});

			// Show the output in a message
			window.showInformationMessage(comment)

		} catch (error) {
			console.error(error);
			return displayError("unableToConnect");
		}
		
	}));

	context.subscriptions.push(vscode.commands.registerCommand('gpt-coder.translateQuery', async function () {
		// The code you place here will be executed every time your command is executed
		const editor = window.activeTextEditor;
		if (!editor) return displayError("noEditor");

		const language1 = await vscode.window.showInputBox({
			placeHolder: "From language",
			prompt: "Query language"
		  });
		
		if (!language1.length) return displayError("Please enter the query language");

		const language2 = await vscode.window.showInputBox({
			placeHolder: "To language",
			prompt: "To query language"
		});

		if (!language2.length) return displayError("Please enter the language");

		const query = await vscode.window.showInputBox({
			placeHolder: "Query",
			prompt: "Query to Translate"
		});

		if (!query.length) return displayError("Please enter the query");
		

		// Aquire the selected text
		const selection = editor.selection;
		// const selectedText = editor.document.getText(selection);
		

		const lang = editor.document.languageId;

		try {
			const comment = styleAsComment(await getChatGPTAnswer(getPromptForTranslate(language1,language2, query)), lang);
			// Replace the selected text with the output
			editor.edit((editBuilder) => {
				editBuilder.insert(selection.end, `\n${comment}`);
			});

			// Show the output in a message
			window.showInformationMessage(comment)

		} catch (error) {
			console.error(error);
			return displayError("unableToConnect");
		}
		
	}));



	const provider = new ColorsViewProvider(context.extensionUri);

    context.subscriptions.push(vscode.window.registerWebviewViewProvider(ColorsViewProvider.viewType, provider));

}

class ColorsViewProvider {

    constructor(_extensionUri) {
        this._extensionUri = _extensionUri;
    }

    async resolveWebviewView(webviewView, context, _token) {
        this._view = webviewView;
        webviewView.webview.options = {
            // Allow scripts in the webview
            enableScripts: true,
            localResourceRoots: [
                this._extensionUri
            ]
        };
        webviewView.webview.html = this._getHtmlForWebview(webviewView.webview);
        webviewView.webview.onDidReceiveMessage(async (data) => {
            switch (data.type) {
				case 'getQuery':
					{	
						// vscode.window.activeTextEditor?.insert(selection.end, new vscode.SnippetString(`#${data.value}`));
						if(!data.language.trim()){
							vscode.window.showErrorMessage("Please enter target lanuage");
						}
						else if(!data.prompt.trim()){
							vscode.window.showErrorMessage("Please enter a prompt");
						}
						else{
							const editor = vscode.window.activeTextEditor;
							const selection = editor.selection;
							const lang = editor.document.languageId;
							if (!editor) return displayError("noEditor");
							try{
								const comment = styleAsComment(await getChatGPTAnswer(getPromptForQuery(data.language, data.prompt)), lang);
							}catch(err){
								console.log(err)
							}
							
							editor.edit((editBuilder) => {
								editBuilder.insert(selection.end, `\n${comment}`);
							});
							window.showInformationMessage(comment)
						}
						break;
					}
				case 'explainQuery':
					{	
						// vscode.window.activeTextEditor?.insert(selection.end, new vscode.SnippetString(`#${data.value}`));
						if(!data.language.trim()){
							vscode.window.showErrorMessage("Please enter lanuage");
						}
						else if(!data.query.trim()){
							vscode.window.showErrorMessage("Please enter a query");
						}
						else{
							const editor = vscode.window.activeTextEditor;
							const selection = editor.selection;
							const lang = editor.document.languageId;
							if (!editor) return displayError("noEditor");
							const comment = styleAsComment(await getChatGPTAnswer(getPromptForQueryExplain(data.language, data.query)), lang);
							editor.edit((editBuilder) => {
								editBuilder.insert(selection.end, `\n${comment}`);
							});
							window.showInformationMessage(comment)
						}
						break;
					}
				case 'queryTranslate':
					{	
						// vscode.window.activeTextEditor?.insert(selection.end, new vscode.SnippetString(`#${data.value}`));
						if(!data.language1.trim()){
							vscode.window.showErrorMessage("Please enter source lanuage");
						}
						if(!data.language2.trim()){
							vscode.window.showErrorMessage("Please enter target lanuage");
						}
						else if(!data.prompt.trim()){
							vscode.window.showErrorMessage("Please enter a prompt");
						}
						else{
							const editor = vscode.window.activeTextEditor;
							const selection = editor.selection;
							const lang = editor.document.languageId;
							if (!editor) return displayError("noEditor");
							const comment = styleAsComment(await getChatGPTAnswer(getPromptForTranslate(data.language1, data.language2, data.prompt)), lang);
							editor.edit((editBuilder) => {
								editBuilder.insert(selection.end, `\n${comment}`);
							});
							window.showInformationMessage(comment)
						}
						break;
					}
				case 'showError':
					{	
						vscode.window.showErrorMessage(data.value);
						break;
					}
            }
        });
    }


    _getHtmlForWebview(webview) {
        // Get the local path to main script run in the webview, then convert it to a uri we can use in the webview.
        const scriptUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'dist', 'main.js'));
		const scriptUri2 = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'media', 'main.js'));
        // Do the same for the stylesheet.
		const styleMainUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'dist', 'main.css'));
        const styleResetUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'media', 'reset.css'));
        const styleVSCodeUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'media', 'vscode.css'));
        // Use a nonce to only allow a specific script to be run.
        const nonce = getNonce();
        return `<!DOCTYPE html>
			<html lang="en">
			<head>
				<meta charset="UTF-8">

				<!--
					Use a content security policy to only allow loading styles from our extension directory,
					and only allow scripts that have a specific nonce.
					(See the 'webview-sample' extension sample for img-src content security policy examples)
				-->
				<meta http-equiv="Content-Security-Policy" content="default-src 'none'; style-src ${webview.cspSource}; script-src 'nonce-${nonce}';">

				<meta name="viewport" content="width=device-width, initial-scale=1.0">

				<link href="${styleResetUri}" rel="stylesheet">
				<link href="${styleVSCodeUri}" rel="stylesheet">
				<link href="${styleMainUri}" rel="stylesheet">

				<title>GPT CODER</title>
			</head>
			<body>

				<div id="app"></div>

				<script nonce="${nonce}" src="${scriptUri}"></script>
				<script nonce="${nonce}" src="${scriptUri2}"></script>
			</body>
			</html>`;
    }
}


ColorsViewProvider.viewType = 'gpt-coder.colorsView';
function getNonce() {
    let text = '';
    const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    for (let i = 0; i < 32; i++) {
        text += possible.charAt(Math.floor(Math.random() * possible.length));
    }
    return text;
}

function getPromptForQuery(prompt, language='sql'){
	return `generate ${language} query for ${prompt}`;
}

function getPromptForTranslate(from, to, query){
	return `generate ${to} equivalent query of ${from} query "${query}"`;
}
function getPromptForQueryExplain(language, query){
	return `explain ${language} query "${query}"`;
}
async function getChatGPTAnswer(prompt){
	const response = await openai.createCompletion({
		model: "text-davinci-003",
		prompt: `${prompt}`,
		temperature: 0,
		max_tokens: 1000,
		top_p: 1,
		frequency_penalty: 0.5,
		presence_penalty: 0,
	});

	if (!response.data) return displayError("serverSentNothing");

	// Get the output from the API response
	return response.data.choices[0].text.trim();
}

// This method is called when your extension is deactivated
function deactivate() { }

module.exports = {
	activate,
	deactivate
}
